File: app/src/main/java/com/vwoom/timelapsecollection/notification/NotificationAlarm.java
package com.vwoom.timelapsecollection.notification;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.PowerManager;
import android.util.Log;

import com.vwoom.timelapsecollection.utils.Keys;

import java.util.Calendar;

public class NotificationAlarm extends BroadcastReceiver {

    private static final String TAG = NotificationAlarm.class.getSimpleName();

    /* This receives the pending intent broadcast and creates the notification */
    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d(TAG, "Notification Tracker: Receiving intent to create notification");
        // Set up the wake lock
        PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        PowerManager.WakeLock w1 = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG);
        long timeout = 1000;    // release wakelock after 1 second
        w1.acquire(timeout);

        // Get the request code
        int requestCode = intent.getIntExtra(Keys.REQUEST_CODE, 0);

        Log.d(TAG, "Notification Tracker: Request code is " + requestCode);
        Log.d(TAG, "Notification Tracker: Action type is " + intent.getAction());

        // Send the notification
        NotificationUtils.notifyUserOfScheduledProjects(context, requestCode);

        // Release the wake lock
        w1.release();
    }

    /* Set an alarm for tomorrow */
    public void setAlarmForTomorrow(Context context){
        Log.d(TAG, "Notification Tracker: Setting alarm for tomorrow");

        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());

        int today = calendar.get(Calendar.DAY_OF_YEAR);
        int requestCode = today + 1;

        Log.d(TAG, "Notification Tracker: Request code / day of year is " + requestCode);
        AlarmManager alarmManager = (AlarmManager)context.getSystemService(Context.ALARM_SERVICE);

        // Create the intent
        Intent intent = new Intent(context, NotificationAlarm.class);
        intent.putExtra(Keys.REQUEST_CODE, requestCode);

        intent.setAction(Keys.CREATE_NOTIFICATION_AUTHORITY + requestCode);

        PendingIntent pendingIntent = PendingIntent
                .getBroadcast(context, requestCode, intent, PendingIntent.FLAG_UPDATE_CURRENT);

        // Convert tomorrow into a timestamp at a set notification time
        long timestampForTomorrow = NotificationUtils.convertDayOfYearToNotificationTime(requestCode);

        alarmManager.set(AlarmManager.RTC_WAKEUP, timestampForTomorrow, pendingIntent);
    }

    /* Cancels an alarm by the day of year */
    public void cancelAlarmByDay(Context context, int dayOfYear){
        Log.d(TAG, "Notification Tracker: Canceling alarm for day " + dayOfYear);
        // Set up the alarm manager
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);

        // Create the intent
        Intent intent = new Intent(context, NotificationAlarm.class);
        // Note: intent action is modified by project id same as above
        intent.setAction(Keys.CREATE_NOTIFICATION_AUTHORITY + dayOfYear);

        // Create the pending intent identified by both request code and intent
        PendingIntent sender = PendingIntent.getBroadcast(context, dayOfYear, intent, 0);

        // Now cancel the alarm
        alarmManager.cancel(sender);
    }

    /* Cancels any alarms set for today or tomorrow */
    public void cancelAlarms(Context context){
        Log.d(TAG, "Notification Tracker: Canceling alarms for today and tomorrow");
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());

        int today = calendar.get(Calendar.DAY_OF_YEAR);
        int tomorrow = today + 1;

        cancelAlarmByDay(context, today);
        cancelAlarmByDay(context, tomorrow);
    }
}

File: app/src/main/java/com/vwoom/timelapsegallery/notification/NotificationAlarm.kt
package com.vwoom.timelapsegallery.notification

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.PowerManager
import com.vwoom.timelapsegallery.BuildConfig
import com.vwoom.timelapsegallery.data.repository.ProjectRepository
import com.vwoom.timelapsegallery.notification.NotificationUtils.clearPreviousNotifications
import com.vwoom.timelapsegallery.notification.NotificationUtils.convertDayOfYearToNotificationTime
import com.vwoom.timelapsegallery.notification.NotificationUtils.notifyUserOfScheduledProjects
import com.vwoom.timelapsegallery.di.InjectorUtils
import com.vwoom.timelapsegallery.utils.ProjectUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import timber.log.Timber
import java.util.*

class NotificationAlarm : BroadcastReceiver() {

    private lateinit var projectRepository: ProjectRepository

    // This receives the pending intent broadcast from the notification alarm and creates a notification
    override fun onReceive(context: Context, intent: Intent) {
        Timber.d("Notification Tracker: Receiving intent to create notification")

        // Set up the wake lock
        val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        val w1 = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG)
        val timeout: Long = 1000 // release wakelock after 1 second
        w1.acquire(timeout)

        // Get the request code which is the day of the year
        val requestCode = intent.getIntExtra(REQUEST_CODE, 0)
        Timber.d("Notification Tracker: Request code is $requestCode")
        Timber.d("Notification Tracker: Action type is ${intent.action}")
        // Clear any previous notifications that are unused
        clearPreviousNotifications(context)

        GlobalScope.launch(Dispatchers.IO) {
            // Get the scheduled projects and figure out if any projects are due today
            projectRepository = InjectorUtils.getProjectRepository(context)
            val scheduledProjects = projectRepository.getScheduledProjectViews()
            val dueProjects = scheduledProjects.filter {
                ProjectUtils.isProjectDueToday(it)
            }

            Timber.d("there are ${dueProjects.size} projects due")
            Timber.d("due projects is not empty ${dueProjects.isNotEmpty()}")
            // If there are projects due today, send the notification
            if (dueProjects.isNotEmpty()) notifyUserOfScheduledProjects(context, requestCode)
            // Release the wake lock
            w1.release()
        }
    }

    // Sets an alarm for tomorrow at the shared preference time
    fun setAlarmForTomorrow(context: Context) {
        Timber.d("Notification Tracker: Setting alarm for tomorrow")
        // Get the day of the year and set it to the request code
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = System.currentTimeMillis()
        val today = calendar[Calendar.DAY_OF_YEAR]
        val requestCode = today + 1

        // Create the pending intent to broadcast on that day
        Timber.d("Notification Tracker: Request code / day of year is $requestCode")
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, NotificationAlarm::class.java)
        intent.putExtra(REQUEST_CODE, requestCode)
        intent.action = CREATE_NOTIFICATION_AUTHORITY + requestCode
        val pendingIntent = PendingIntent
                .getBroadcast(context, requestCode, intent, PendingIntent.FLAG_UPDATE_CURRENT)

        // Get the timestamp from the shared preference for the day at that notification time
        val timestampForTomorrow = convertDayOfYearToNotificationTime(requestCode, context)

        // Set the alarm
        alarmManager[AlarmManager.RTC_WAKEUP, timestampForTomorrow] = pendingIntent
    }

    // Helper function that Cancels an alarm by the day of year
    private fun cancelAlarmByDay(context: Context, dayOfYear: Int) {
        Timber.d("Notification Tracker: Canceling alarm for day $dayOfYear")

        // Identify the alarm
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, NotificationAlarm::class.java)
        intent.action = CREATE_NOTIFICATION_AUTHORITY + dayOfYear
        // Create the pending intent to remove the alarm by day
        val sender = PendingIntent.getBroadcast(context, dayOfYear, intent, 0)

        // Now cancel the alarm
        alarmManager.cancel(sender)
    }

    // Cancels any alarms set for today or tomorrow
    // This should be any alarms set
    fun cancelAlarms(context: Context) {
        Timber.d("Notification Tracker: Canceling alarms for today and tomorrow")
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = System.currentTimeMillis()
        val today = calendar[Calendar.DAY_OF_YEAR]
        val tomorrow = today + 1
        cancelAlarmByDay(context, today)
        cancelAlarmByDay(context, tomorrow)
    }

    companion object {
        private val TAG = NotificationAlarm::class.java.simpleName
        private const val REQUEST_CODE = "request_code"
        private const val applicationId = BuildConfig.APPLICATION_ID
        private const val CREATE_NOTIFICATION_AUTHORITY = "$applicationId.CREATE_NOTIFICATION"
    }
}